### DISCUSSION TITLE
checksum.c:127:5: ошибка: «struct tcphdr» не содержит элемента с именем «th_sum»

### DISCUSSION THREAD

**Discussion by v-abramoff** [bol-van/zapret#308] [Q&A]
Привет,

Пытаюсь собрать под Centos 7 - сборка завершается ошибкой:

```
make[1]: Вход в каталог `/home/abramoff/zapret/nfq'
cc -s -std=gnu99 -Wno-logical-op-parentheses -O3 -o nfqws *.c crypto/*.c  -lnetfilter_queue -lnfnetlink -lz
checksum.c: В функции «tcp4_fix_checksum»:
checksum.c:127:5: ошибка: «struct tcphdr» не содержит элемента с именем «th_sum»
  tcp->th_sum = 0;
```

И далее много подобных ошибок. Что можно предпринять?

**Comment by bol-van** [comment]
Должно быть сильно устаревший хедер /usr/include/netinet/tcp.h
У меня есть старенький iso от centos 7. Но там уже все репы протухли
даже не посмотреть

если нет возможности апгрейднуть ос с этого мамонта, стоит использовать готовые статические бинарики
или собрать их самому с CFLAGS=-static на системе с musl. alpine linux подойдет

запостите сюда ваш /usr/include/netinet/tcp.h
посмотрю. либо там вообще нет нужного, либо какого-то define не хватает

**Comment by v-abramoff** [comment]
> Должно быть сильно устаревший хедер /usr/include/netinet/tcp.h

Там вот такой код (также см. вложение):
[tcp.zip](https://github.com/user-attachments/files/16676011/tcp.zip)


```
# ifdef __FAVOR_BSD
typedef u_int32_t tcp_seq;
/*
 * TCP header.
 * Per RFC 793, September, 1981.
 */
struct tcphdr
  {
    u_int16_t th_sport;         /* source port */
    u_int16_t th_dport;         /* destination port */
    tcp_seq th_seq;             /* sequence number */
    tcp_seq th_ack;             /* acknowledgement number */
#  if __BYTE_ORDER == __LITTLE_ENDIAN
    u_int8_t th_x2:4;           /* (unused) */
    u_int8_t th_off:4;          /* data offset */
#  endif
#  if __BYTE_ORDER == __BIG_ENDIAN
    u_int8_t th_off:4;          /* data offset */
    u_int8_t th_x2:4;           /* (unused) */
#  endif
    u_int8_t th_flags;
#  define TH_FIN        0x01
#  define TH_SYN        0x02
#  define TH_RST        0x04
#  define TH_PUSH       0x08
#  define TH_ACK        0x10
#  define TH_URG        0x20
    u_int16_t th_win;           /* window */
    u_int16_t th_sum;           /* checksum */
    u_int16_t th_urp;           /* urgent pointer */
};

# else /* !__FAVOR_BSD */
struct tcphdr
  {
    u_int16_t source;
    u_int16_t dest;
    u_int32_t seq;
    u_int32_t ack_seq;
#  if __BYTE_ORDER == __LITTLE_ENDIAN
    u_int16_t res1:4;
    u_int16_t doff:4;
    u_int16_t fin:1;
    u_int16_t syn:1;
    u_int16_t rst:1;
    u_int16_t psh:1;
    u_int16_t ack:1;
    u_int16_t urg:1;
    u_int16_t res2:2;
#  elif __BYTE_ORDER == __BIG_ENDIAN
    u_int16_t doff:4;
    u_int16_t res1:4;
    u_int16_t res2:2;
    u_int16_t urg:1;
    u_int16_t ack:1;
    u_int16_t psh:1;
    u_int16_t rst:1;
    u_int16_t syn:1;
    u_int16_t fin:1;
#  else
#   error "Adjust your <bits/endian.h> defines"
#  endif
    u_int16_t window;
    u_int16_t check;
    u_int16_t urg_ptr;
};
# endif /* __FAVOR_BSD */
```

Что такое __FAVOR_BSD - я не понимаю.

>У меня есть старенький iso от centos 7. Но там уже все репы протухли

Репозитории есть на vault.centos.org

**Comment by bol-van** [comment]
Ясно, Проблема в favor bsd
В tcp.h есть старая линух-проприетарная версия tcphdr
Но дальше приняли за основу определение tcphdr из BSD
В новых tcp.h уже этого дефайна нет

Придется значит задефайнить

Репы пробовал менять,но yum то на одном виснет, то на другом. дисаблил плугины - еще на другом повисло
забил, пробуем так пропатчит

**Comment by bol-van** [comment]
пробуйте последний комит

**Reply by v-abramoff** [reply]
Ничего не поменялось. Те же ошибки.

**Reply by bol-van** [reply]
проблемы с yum были из-за ipv6. vaults не отвечает на ipv6
ладно, с этим я разобрался, накатил gcc и прочее

теперь должно собираться

**Reply by v-abramoff** [reply]
Благодарю, теперь ошибок нет!
